src/async_internal.rs

```diff
@@ -0,0 +1,230 @@
+///
+#[async_trait]
+pub trait AsyncPasswordTokenRequest<'a, TE, TR, TT>
+where
+    TE: ErrorResponse + 'static,
+    TR: TokenResponse<TT> + Send,
+    TT: TokenType + Send,
+{
+    ///
+    /// Asynchronously sends the request to the authorization server and awaits a response.
+    ///
+    async fn request_async<C, F, RE>(self, http_client: C) -> Result<TR, RequestTokenError<RE, TE>>
+    where
+        C: FnOnce(HttpRequest) -> F + Send,
+        F: Future<Output = Result<HttpResponse, RE>> + Send,
+        RE: Fail;
+}
+
+#[async_trait]
+impl<'a, TE, TR, TT> AsyncPasswordTokenRequest<'a, TE, TR, TT>
+    for PasswordTokenRequest<'a, TE, TR, TT>
+where
+    TE: ErrorResponse + 'static,
+    TR: TokenResponse<TT> + Send,
+    TT: TokenType + Send,
+{
+    ///
+    /// Asynchronously sends the request to the authorization server and awaits a response.
+    ///
+    async fn request_async<C, F, RE>(self, http_client: C) -> Result<TR, RequestTokenError<RE, TE>>
+    where
+        C: FnOnce(HttpRequest) -> F + Send,
+        F: Future<Output = Result<HttpResponse, RE>> + Send,
+        RE: Fail,
+    {
+        let http_request = self.prepare_request()?;
+        let http_response = http_client(http_request)
+            .await
+            .map_err(RequestTokenError::Request)?;
+        token_response(http_response)
+    }
+}
+
+///
+/// Asynchronous request to exchange client credentials for an access token.
+///
+#[async_trait]
+pub trait AsyncClientCredentialsTokenRequest<'a, TE, TR, TT>
+where
+    TE: ErrorResponse + 'static,
+    TR: TokenResponse<TT> + Send,
+    TT: TokenType + Send,
+{
+    ///
+    /// Asynchronously sends the request to the authorization server and awaits a response.
+    ///
+    async fn request_async<C, F, RE>(self, http_client: C) -> Result<TR, RequestTokenError<RE, TE>>
+    where
+        C: FnOnce(HttpRequest) -> F + Send,
+        F: Future<Output = Result<HttpResponse, RE>> + Send,
+        RE: Fail;
+}
+
+#[async_trait]
+impl<'a, TE, TR, TT> AsyncClientCredentialsTokenRequest<'a, TE, TR, TT>
+    for ClientCredentialsTokenRequest<'a, TE, TR, TT>
+where
+    TE: ErrorResponse + 'static,
+    TR: TokenResponse<TT> + Send,
+    TT: TokenType + Send,
+{
+    ///
+    /// Asynchronously sends the request to the authorization server and awaits a response.
+    ///
+    async fn request_async<C, F, RE>(self, http_client: C) -> Result<TR, RequestTokenError<RE, TE>>
+    where
+        C: FnOnce(HttpRequest) -> F + Send,
+        F: Future<Output = Result<HttpResponse, RE>> + Send,
+        RE: Fail,
+    {
+        let http_request = self.prepare_request()?;
+        let http_response = http_client(http_request)
+            .await
+            .map_err(RequestTokenError::Request)?;
+        token_response(http_response)
+    }
+}
+
+///
+/// Asynchronous HTTP client.
+///
+pub async fn async_http_client(request: HttpRequest) -> Result<HttpResponse, Error> {
```

Comment:
would it be possible to move this to the `reqwest` module? I think we can do a similar trick and put this in a private module inside `oauth2::reqwest` and then conditionally `pub use` it.

to avoid the name conflict, maybe we should rename the old async interface to `future_http_client` and each of the old `request_async` interfaces to `request_future`? that should allow them to coexist, I think, and fix the `--all-features` issue. it should also make the library easier to use since users can specify any combination of features without running into compilation errors.